compiler/rustc_lint/src/types.rs

```diff
@@ -851,12 +851,13 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
         use FfiResult::*;
 
         if def.repr.transparent() {
-            // Can assume that only one field is not a ZST, so only check
+            // Can assume that at most one field is not a ZST, so only check
             // that field's type for FFI-safety.
             if let Some(field) = transparent_newtype_field(self.cx.tcx, variant) {
                 self.check_field_type_for_ffi(cache, field, substs)
             } else {
-                bug!("malformed transparent type");
+                // All fields are ZSTs
+                FfiSafe
```

Comment:
I'm not sure if this is the correct behavior; in the implementation I had started, I initially went with `FfiSafe` but then switched to `FfiUnsafe`. The reason is that according to the issue where the new behavior was discussed (https://github.com/rust-lang/rust/issues/77841#issuecomment-716575747) \[emphasis mine\],

> That's a good point. So here is a proposed set of rules:
> 
>   * A `repr(transparent)` type `T` must meet the following rules:
>       
>     * It may have any number of 1-ZST fields
>     * In addition, it may have at most one other field of type U
> 
> 
> If that other field is present, then T behaves like type U for ABI purposes. **Otherwise, T consists only of 1-ZST types, and it behaves like `()`, which is the "canonical" 1-ZST.**

and `()` is currently FFI-**unsafe**:

```
warning: `extern` block uses type `()`, which is not FFI-safe
 --> src/lib.rs:2:18
  |
2 |     fn good17(p: ());
  |                  ^^ not FFI-safe
  |
  = note: `#[warn(improper_ctypes)]` on by default
  = help: consider using a struct instead
  = note: tuples have unspecified layout
```

There's also this later comment in that thread (https://github.com/rust-lang/rust/issues/77841#issuecomment-716698421):

> Hmm, the lint considers `()` an improper_ctype. But then an empty struct is UB(!?) in C, so I guess there isn't a canonical ABI for this.

Reply:
I think this is correct -- no fields ought to behave like `()`